docs/resources/server/advancements.md

```diff
@@ -92,48 +80,54 @@ public boolean matches(ItemStack stack) {
-  // Read conditions from JSON: item
-  return new ExampleTriggerInstance(player, item);
-}
-```
+The `SimpleCriterionTrigger<T>` subclass is responsible for specifying a codec to [serialize] the trigger instance `T` and supplying a method to check trigger instances and run attached listeners on success.
 
-Finally, a method is defined to check all trigger instances and run the listeners if their condition is met. This method takes in the `ServerPlayer` and whatever other data defined by the matching method in the `AbstractCriterionTriggerInstance` subclass. This method should internally call `SimpleCriterionTrigger#trigger` to properly handle checking all listeners. Most trigger instances call this method `#trigger`.
+The later is done by defining a method to check all trigger instances and run the listeners if their condition is met. This method takes in the `ServerPlayer` and whatever other data defined by the matching method in the `SimpleCriterionTrigger.SimpleInstance` subclass. This method should internally call `SimpleCriterionTrigger#trigger` to properly handle checking all listeners. Most trigger instances call this method `#trigger`.
 
 ```java
-// This method is unique for each trigger and is as such not overridden
+// This method is unique for each trigger and is as such not a method to override
 public void trigger(ServerPlayer player, ItemStack stack) {
   this.trigger(player,
-    // The condition checker method within the AbstractCriterionTriggerInstance subclass
+    // The condition checker method within the SimpleCriterionTrigger.SimpleInstance subclass
     triggerInstance -> triggerInstance.matches(stack)
   );
 }
 ```
 
-Afterwards, an instance should be registered using `CriteriaTriggers#register` during `FMLCommonSetupEvent`.
+Finally, instances must be registered on the `Registries.TRIGGER_TYPE` registry. Techniques for doing so can be found under [Registries][registration].
 
-:::danger
-`CriteriaTriggers#register` must be enqueued to the synchronous work queue via `FMLCommonSetupEvent#enqueueWork` as the method is not thread-safe.
-:::
+### Serialization
+
+A [codec] must be defined to serialize and deserialize the trigger instance. Vanilla typically creates this codec as a constant within the instance implementation.
+
+```java
+// in ExampleTriggerInstance
+public static final Codec<ExampleTriggerInstance> CODEC = RecordCodecBuilder.create(instance -> instance.group(
+  ExtraCodecs.strictOptionalField(EntityPredicate.ADVANCEMENT_CODEC, "player").forGetter(ExampleTriggerInstance::player),
+  ItemPredicate.CODEC.fieldOf("item").forGetter(ExampleTriggerInstance::item)
+).apply(instance, ExampleTriggerInstance::new));
+```
+
+This is then returned by the trigger's `#codec` method.
+
+```java
+// in ExampleTrigger
+@Override
+public Codec<ExampleTriggerInstance> codec() {
+  return ExampleTriggerInstance.CODEC;
+}
```

Comment:
I think just for ease of understanding, the structure should be something like this with the comments:

```java
class ExampleTrigger<Instance> {
    codec() { Instance.CODEC; }

    class Instance {
        CODEC = ...;
    }
}
```

Reply:
Agreed